File: 208. Implement Trie (Prefix Tree).py
# 208. Implement Trie (Prefix Tree).            -Medium-

# Topic:Hash Table, String, Design, Trie.

"""
### Task:
---
A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve 
keys in a dataset of strings. There are various applications of this data structure, such as autocomplete and spellchecker.

Implement the Trie class:

- Trie() Initializes the trie object.
- void insert(String word) Inserts the string word into the trie.
- boolean search(String word) Returns true if the string word is in the trie (i.e., was inserted before), and false otherwise.
- boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the prefix prefix, 
  and false otherwise.
 
Example 1:
Input
["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
Output
[null, null, true, false, true, null, true]

Explanation
Trie trie = new Trie();
trie.insert("apple");
trie.search("apple");   // return True
trie.search("app");     // return False
trie.startsWith("app"); // return True
trie.insert("app");
trie.search("app");     // return True

Constraints:
1 <= word.length, prefix.length <= 2000
word and prefix consist only of lowercase English letters.
At most 3 * 10^4 calls in total will be made to insert, search, and startsWith.


### Testcase:
---
["Trie","insert","search","search","startsWith","insert","search"]
[[],["apple"],["apple"],["app"],["app"],["app"],["app"]]

### Code:
class Trie:

    def __init__(self):
        

    def insert(self, word: str) -> None:
        

    def search(self, word: str) -> bool:
        

    def startsWith(self, prefix: str) -> bool:
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

"""
### Solution: ------------------------------------

class TrieNode:
    def __init__(self):
        # Each node stores its children in a dictionary.
        # Keys are characters, and values are TrieNode objects.
        self.children = {}
        # Flag to indicate if a word ends at this node.
        self.isEndOfWord = False

class Trie:
    def __init__(self):
        # The root node of the Trie, which is an empty node initially.
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        # Start from the root node.
        node = self.root
        for char in word:
            # For each character in the word, find the corresponding child node.
            if char not in node.children:
                # If the child node doesn't exist, create a new TrieNode.
                node.children[char] = TrieNode()
            # Move to the child node.
            node = node.children[char]
        # After the last character, mark this node as the end of a word.
        node.isEndOfWord = True

    def search(self, word: str) -> bool:
        # Start from the root node.
        node = self.root
        for char in word:
            # For each character in the word, find the corresponding child node.
            if char not in node.children:
                # If a character is not found, the word does not exist in the Trie.
                return False
            # Move to the child node.
            node = node.children[char]
        # Return true if the current node is marked as the end of the word.
        return node.isEndOfWord

    def startsWith(self, prefix: str) -> bool:
        # Start from the root node.
        node = self.root
        for char in prefix:
            # For each character in the prefix, find the corresponding child node.
            if char not in node.children:
                # If a character is not found, no words start with this prefix.
                return False
            # Move to the child node.
            node = node.children[char

]
        # If the prefix is traversed successfully, return true.
        # It's not necessary for the node to be the end of a word; we only care about the prefix.
        return True
    

### Description: =================================

'''
### Comments Explanation:
- **TrieNode Class**: A helper class representing each node in the Trie. Each node has a `children` dictionary, mapping characters to their corresponding child nodes, and a boolean `isEndOfWord` to indicate if a node is the end of a word.
- **Trie Class Initialization**: In the `__init__` method of the `Trie` class, a root node is created. This root node is essentially an empty node to start building the Trie.
- **Insert Method**: This method iterates through each character of the input word. For each character, it checks if the corresponding child node exists in the current node's `children`. If not, it creates a new `TrieNode`. It then moves to this child node and continues. Finally, it marks the node corresponding to the last character as the end of a word.
- **Search Method**: Similar to `insert`, but instead of creating new nodes, it checks if each character of the word exists in the Trie. If any character is missing, it returns `False`. If all characters are found and the final node is marked as the end of a word, it returns `True`.
- **StartsWith Method**: This method checks if a prefix exists in the Trie. It is similar to `search`, but it only needs to check if the nodes corresponding to each character in the prefix exist. It does not need to check if the final node is marked as the end of a word.

These comments provide clear explanations of each part of the Trie implementation, detailing the functionality and purpose of each method and the `TrieNode` class, which together facilitate efficient string insertion, search, and prefix matching.

Implementing a Trie (or Prefix Tree) is a classic problem that tests understanding of tree-like data structures and their applications in operations like search and prefix matching. Let's dive into the solution.

### Problem Analysis:
A Trie is a specialized tree used in searching, most often with strings. The key tasks are:
- **Insertion**: Add a word to the Trie.
- **Search**: Check if a word is present in the Trie.
- **Prefix Matching**: Check if any word in the Trie starts with a given prefix.

### Approach:
1. **Trie Node Structure**:
   - Each node should contain a map (or array) pointing to its child nodes and a boolean to mark the end of a word.

2. **Insertion**:
   - Start from the root and for each character in the word, go to the corresponding child node.
   - If the child node doesn’t exist, create a new node.
   - Mark the end of the word in the last node.

3. **Search**:
   - Similar to insertion, but check if each character of the word is present in Trie.
   - If any character is missing, return false.
   - If all characters are found and the last node is marked as the end of a word, return true.

4. **Starts With (Prefix Matching)**:
   - Similar to search, but it's only necessary to check if the prefix is present, not whether it marks the end of a word.


### Explanation:
- **TrieNode Class**: Represents a node in the Trie. Each node has a map (`children`) to represent its children and a flag (`isEndOfWord`) to indicate if it's the end

 of a word.
- **Trie Class Initialization**: Initializes the Trie with a root node.
- **Insert Method**: Iterates through each character of the word. For each character, if the corresponding child node does not exist, it creates a new one. Finally, it marks the last node as the end of a word.
- **Search Method**: Similar to insert, it traverses the Trie for each character in the word. If any character node is missing, it returns false. It also checks if the last node is marked as the end of a word.
- **StartsWith Method**: Traverses the Trie for each character in the prefix. If any character node is missing, it returns false. Unlike search, it doesn't need to check if the last node is the end of a word, as it's only concerned with the prefix.

### Time Complexity:
- **Insert**: O(m), where m is the length of the word.
- **Search**: O(m), for a word of length m.
- **StartsWith**: O(p), for a prefix of length p.

### Test Case:
For the test case `["Trie", "insert", "search", "search", "startsWith", "insert", "search"]` and `[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]`:
- Create a Trie.
- Insert "apple".
- Search "apple" -> returns `True`.
- Search "app" -> returns `False` (as "app" is not a complete word in the Trie).
- Check startsWith "app" -> returns `True`.
- Insert "app".
- Search "app" -> returns `True`.

This Trie implementation efficiently handles the operations of insertion, search, and prefix matching in a dataset of strings, making it suitable for applications like autocomplete and spellcheckers.

'''

File: 208. Implement Trie (Prefix Tree).rs
// 208. Implement Trie (Prefix Tree).           -medium-

// Topic: Hash Table, String, Design, Trie.


/*
### Task:
---
A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve 
keys in a dataset of strings. There are various applications of this data structure, such as autocomplete and spellchecker.

Implement the Trie class:

- Trie() Initializes the trie object.
- void insert(String word) Inserts the string word into the trie.
- boolean search(String word) Returns true if the string word is in the trie (i.e., was inserted before), and false otherwise.
- boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the prefix prefix, 
  and false otherwise.
 
Example 1:
Input
["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
Output
[null, null, true, false, true, null, true]

Explanation
Trie trie = new Trie();
trie.insert("apple");
trie.search("apple");   // return True
trie.search("app");     // return False
trie.startsWith("app"); // return True
trie.insert("app");
trie.search("app");     // return True

Constraints:
1 <= word.length, prefix.length <= 2000
word and prefix consist only of lowercase English letters.
At most 3 * 10^4 calls in total will be made to insert, search, and startsWith.


### Testcase:
---
["Trie","insert","search","search","startsWith","insert","search"]
[[],["apple"],["apple"],["app"],["app"],["app"],["app"]]

### Code:
struct Trie {

}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Trie {

    fn new() -> Self {
        
    }
    
    fn insert(&self, word: String) {
        
    }
    
    fn search(&self, word: String) -> bool {
        
    }
    
    fn starts_with(&self, prefix: String) -> bool {
        
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * let obj = Trie::new();
 * obj.insert(word);
 * let ret_2: bool = obj.search(word);
 * let ret_3: bool = obj.starts_with(prefix);
 */
*/

// Solution: ------------------------------------


// Constant representing the size of the alphabet (26 English lowercase letters).
const ALPHABET_SIZE: usize = 26;

// Definition of a TrieNode struct.
struct TrieNode {
    // Each node has `ALPHABET_SIZE` children, stored in an array.
    // `Option<Box<TrieNode>>` is used for each child, where `None` indicates no child node.
    children: [Option<Box<TrieNode>>; ALPHABET_SIZE],

    // Flag to mark the end of a word.
    is_end_of_word: bool,
}

impl TrieNode {
    // Constructor for a new TrieNode.
    fn new() -> Self {
        TrieNode {
            // Initialize the children array with `None`, meaning no children initially.
            children: Default::default(),
            is_end_of_word: false,
        }
    }

    // Helper function to map a character to its corresponding index in the children array.
    // Converts a lowercase English letter to a zero-based index.
    fn index(c: char) -> usize {
        (c as usize) - ('a' as usize)
    }
}

// Definition of the Trie struct.
struct Trie {
    // The root node of the Trie.
    root: TrieNode,
}

impl Trie {
    // Constructor for a new Trie.
    fn new() -> Self {
        Trie {
            root: TrieNode::new(),
        }
    }

    // Method to insert a word into the Trie.
    fn insert(&mut self, word: String) {
        let mut node = &mut self.root;
        for c in word.chars() {
            // Determine the index for the character.
            let idx = TrieNode::index(c);
            // Access the child node at the index, creating a new node if it doesn't exist.
            node = node.children[idx].get_or_insert_with(|| Box::new(TrieNode::new()));
        }
        // Mark the end of the word.
        node.is_end_of_word = true;
    }

    // Method to check if a word exists in the Trie.
    fn search(&self, word: String) -> bool {
        let mut node = &self.root;
        for c in word.chars() {
            // Determine the index for the character.
            let idx = TrieNode::index(c);
            // Check if the child node exists; if

 not, return false.
            match node.children[idx].as_ref() {
                Some(n) => node = n,
                None => return false,
            }
        }
        // Return true if the end of the word has been reached, false otherwise.
        node.is_end_of_word
    }

    // Method to check if any word in the Trie starts with the given prefix.
    fn starts_with(&self, prefix: String) -> bool {
        let mut node = &self.root;
        for c in prefix.chars() {
            // Determine the index for the character.
            let idx = TrieNode::index(c);
            // Check if the child node exists; if not, return false.
            match node.children[idx].as_ref() {
                Some(n) => node = n,
                None => return false,
            }
        }
        // If the loop completes without returning false, a prefix match exists.
        true
    }
}



// Description: =================================
/*
### Commentary:

- The `TrieNode` struct represents each node in the Trie. It contains an array of child nodes, with one entry for each letter in the alphabet, and a boolean to indicate whether it represents the end of a word.

- The `Trie` struct represents the entire Trie. It has a root `TrieNode` from which all words are branched out.

- The `insert` method iterates through each character in the input word, creating new nodes as necessary, and marks the end of the word.

- The `search` method checks for the existence of a word in the Trie, traversing node by node according to the word's characters.

- The `starts_with` method verifies if any word in the Trie starts with a given prefix, similar to the `search` method but only checks for the presence of the prefix, not the complete word.

These methods collectively allow the Trie to efficiently store and query strings, particularly useful for tasks like autocomplete and spell checking in large datasets of strings.

### Description of the Trie Data Structure Implementation in Rust

The implementation of the Trie (Prefix Tree) data structure in Rust involves creating a specialized tree-like structure optimized for efficient storage and retrieval of strings. This data structure is particularly useful in applications like autocomplete systems, spell checkers, and other scenarios where quick lookup of words or prefixes is required.



#### Core Components:

1. **TrieNode Structure:**
   - Each node (`TrieNode`) in the Trie represents a single character of a word.
   - It contains an array of child nodes, each corresponding to one of the 26 lowercase English letters

. This array is implemented as `[Option<Box<TrieNode>>; ALPHABET_SIZE]`, where `ALPHABET_SIZE` is 26 for the English alphabet.
   - The `Option<Box<TrieNode>>` type allows each element in the array to either hold a `TrieNode` (boxed, i.e., heap-allocated) or `None`, indicating the absence of a child node for the corresponding letter.
   - A boolean flag `is_end_of_word` indicates whether the node marks the completion of a valid word in the Trie.

2. **Trie Structure:**
   - The Trie itself is represented by the `Trie` struct, which primarily

contains a root `TrieNode`. This root node acts as the entry point to the Trie and does not correspond to any character.
   - The Trie provides a framework for implementing various methods that operate on the TrieNode structure to insert words, search for words, and check for the presence of prefixes.

#### Key Methods:

1. **Insertion (`insert`):**
   - Words are inserted into the Trie character by character.
   - Starting from the root, the method navigates through or creates child nodes for each character in the word.
   - Once the end of the word is reached, the `is_end_of_word` flag is set to `true` on the final node, marking the completion of a word.

2. **Search (`search`):**
   - The search operation checks whether a complete word exists in the Trie.
   - It traverses the Trie from the root, following the path of characters in the word.
   - If the path exists and ends with `is_end_of_word` set to `true`, the word is present in the Trie.

3. **Prefix Checking (`starts_with`):**
   - This operation checks if there is any word in the Trie that starts with a given prefix.
   - Similar to `search`, it traverses the Trie according to the characters in the prefix.
   - The method returns `true` if the path for the entire prefix exists, regardless of the `is_end_of_word` flag.

#### Performance and Use Cases:

- **Efficiency:** The Trie structure is highly efficient for scenarios involving a large number of string lookups. 
   It optimizes both the storage and retrieval of strings, particularly when dealing with large datasets of words.
- **Use Cases:** Commonly used in situations where quick prefix searches are frequent, such as in autocomplete systems, 
   spell checkers, or in implementing dictionaries.

#### Conclusion:

This Rust implementation of the Trie data structure demonstrates an efficient approach to handling strings in a way that 
optimizes for quick insertions and search operations. It leverages Rust's memory safety and efficient data structures, 
like `HashMap`, to provide a robust and performance-oriented solution.


*/
